Add processing model to spec

AloisReitbauer opened the pull request:
<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/trace-context/pull/222.html" title="Last updated on Feb 12, 2019, 4:02 PM UTC (fa851a7)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/trace-context/222/3e0c447...fa851a7.html" title="Last updated on Feb 12, 2019, 4:02 PM UTC (fa851a7)">Diff</a>

AloisReitbauer committed:
added processing model

AloisReitbauer committed:
added processing model to toc

danielkhan committed:
Grammar and wording fixes

- fixed spelling
- edited wording for clarity
- fixed grammar
- added comments and issue links

danielkhan committed:
Revert "Grammar and wording fixes"

This reverts commit 6c320ea3e8090c58005cce060cac17c32988e97f.

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+# Processing Model
+
+This section defines the behaviour of a tracing system when a new request is received by a component. As stated above tracing systems can decide for each individual request whether they want to _forward_ a trace context or _participate_ in a trace context. The processing model for both scenarios will be described individualy and adds a seperate processing model when a tracing system is the first  ...
```
Terminology-wise should it be larger than just "tracing system"? Let's call it "tracing system or platform" to be more inclusive of proxy servers and such

mtwo replied:
Ehhh ... I'm not sure if this distinction matters

AloisReitbauer replied:
Platform is a very generic term I am not sure how this adds value

danielkhan replied:
We should at some point go through the spec and harmonize the terminology. Maybe move all of the explanations into some glossary.
I agree that a clear terminology is key but this should for now not keep us from merging.
Do you agree?

AloisReitbauer replied:
Decided to postpone in cleanup phase later as difference is not substantial

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+# Processing Model
+
+This section defines the behaviour of a tracing system when a new request is received by a component. As stated above tracing systems can decide for each individual request whether they want to _forward_ a trace context or _participate_ in a trace context. The processing model for both scenarios will be described individualy and adds a seperate processing model when a tracing system is the first  ...
```
"and adds" - stylistically it's unclear who adds processing model. Please simplify

danielkhan replied:
This section was rewritten.

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+# Processing Model
+
+This section defines the behaviour of a tracing system when a new request is received by a component. As stated above tracing systems can decide for each individual request whether they want to _forward_ a trace context or _participate_ in a trace context. The processing model for both scenarios will be described individualy and adds a seperate processing model when a tracing system is the first  ...
```
"seperate" => "separate"

danielkhan replied:
This section was rewritten already

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+
+This section defines the behaviour of a tracing system when a new request is received by a component. As stated above tracing systems can decide for each individual request whether they want to _forward_ a trace context or _participate_ in a trace context. The processing model for both scenarios will be described individualy and adds a seperate processing model when a tracing system is the first  ...
+
+## Processing Model in Forwarding Mode
```
should it be Forward processing model? Term "mode" is new here

danielkhan replied:
This section has been rewritten.

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+
+## Processing Model in Forwarding Mode
+
+In forwarding mode a tracing system decides to forward tracing information but to not modify the trace-context headers. The request is then considered as not sampled. test this
```
why `trace-context` used with dashes? There is no dashes in intro to the section

danielkhan replied:
Fixed in update

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+
+## Processing Model in Forwarding Mode
+
+In forwarding mode a tracing system decides to forward tracing information but to not modify the trace-context headers. The request is then considered as not sampled. test this
```
> test this

test what?

SergeyKanzhelev replied:
Should forward be a simple forward? How forward model is related to sampling at all?

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+1. Tracing system receives a `traceparent` header and may receive `tracecontext`. 
+2. Tracing system tries to parse the version of the `traceparent` header.
+     1. If the version prefix cannot be parsed (it's not 2 hex characters followed by
+   dash (`-`)), implementation should restart the trace. *NOTE* : If a system is in forwarding only mode. It cannot start a trace. 
```
What is "If a system is in forwarding only mode."?

mtwo replied:
That's the mode that's being defined here. It defines any system that wants to forward trace headers but will otherwise not participate in the trace.

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+
+1. Tracing system receives a `traceparent` header and may receive `tracecontext`. 
+2. Tracing system tries to parse the version of the `traceparent` header.
+     1. If the version prefix cannot be parsed (it's not 2 hex characters followed by
```
do we need this check when forwarding?

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+    can block certain `tracestate` keys for privacy and security concerns. The
+    second scenario is a truncation of long `tracestate`'s.
+
+## Processing Model for (re)starting a trace
```
stylistically I suggest to split starting and restarting as two separate words. Easier to translate

yurishkuro replied:
Logically this is a subsection of the previous header2.

danielkhan replied:
Fixed in https://github.com/w3c/trace-context/pull/246/commits/85aeacd803375b73125f86e4838ca6b632c39bde

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+## Processing Model for (re)starting a trace
+
+1. **Restarting trace**. All properties - `trace-id`, `parent-id`, `trace-flags`
+   are regenerated. This mutation is used in the services defined as a front
```
"is used" => "can be used" or "is typically used". I don't think this behavior should be limited to a single scenario.

yurishkuro replied:
This mutation is called up in the previous header2 several times, so it is not just for the edge.

danielkhan replied:
This section has be updated in https://github.com/w3c/trace-context/pull/246/commits/25b844e5d4df816153b6147cced330cb6dd34b86 - please review

AloisReitbauer replied:
In case you are restarting a trace you should record the old traceparent. We have not really defined this behaviour

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+
+## Processing Model in Participation Mode
+
+In participation mode the tracing system decides to forward AND modofiy trace-context headers. The request is then consideres as sampled.
```
consideres => considered

danielkhan replied:
Fixed

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+1. Tracing system receives a `traceparent` header and may receive `tracecontext`. 
+2. Tracing system tries to parse the version of the `traceparent` header.
+    1. If the version prefix cannot be parsed (it's not 2 hex characters followed by
+   dash (`-`)), the tracing system creates a new `traceparent` header and removes all entries from `tracecontext`. 
```
> it's not 2 hex characters followed by dash (`-`)

Please do not include specifics of text-based formatting into the semantical processing model. Same processing model should be used for GRPC server

danielkhan replied:
Fixed in https://github.com/w3c/trace-context/pull/246/commits/2037e0bfb3e9a63e304468d6730a35c1121c5d67

SergeyKanzhelev commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,64 @@
+    2. If either `trace-id` or `parent-id` are invalid, the tracing systems does not add the trace context to any outgoing requests.
+*NOTE* : We never explictly defined this, but I think this is what should happen. 
+
+## Processing Model in Participation Mode
```
I think processing mode with only `tracestate` modification should be a valid case. It may be privacy-enforcing proxy or proxy that adds information about itself without touching `traceparent`. Like (potentially) service mesh scenario.

AloisReitbauer replied:
This sounds to me like abusing trace-parent for baggage.

danielkhan replied:
#250

SergeyKanzhelev reviewed (changes requested):
I added some initial comments.

mtwo reviewed (approved):
Sergey had a few grammar nits that I agree with, but this looks good overall

danielkhan committed:
Grammar, Spelling and wording (#225)

* Grammar, Spelling and wording

- Fixed grammar, spelling and edited wording for clarity
- Formatted pending decisions
- Added issue links

* Moved IF conditions to bullets instead of ordered lists

yurishkuro commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,47 @@
+
+In forwarding mode, a tracing system decides to forward tracing information but does not modify the trace-context headers. The request is then considered as not sampled.
+
+1. The tracing system receives a `traceparent` header may receive a `tracecontext`. 
```
Why does the system need to do any parsing in the forwarding mode?

AloisReitbauer replied:
It does not need to, but we had the case where a system parses the headers to drop them if they are invalid. This language was taken from the current version of the spec. 

I am fine to define that in forwarding mode you just blindly forward the header(s)

yurishkuro replied:
I'd say "MAY parse the headers according to ... (participating mode)  and SHOULD/MUST forward the headers unmodified"

AloisReitbauer replied:
What happens if you parse and find that it is invalid?

danielkhan replied:
Let's discuss in https://github.com/w3c/trace-context/issues/251

yurishkuro commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,47 @@
+
+## Processing Model in Participation Mode
+
+In participation mode, the tracing system decides to forward **and** modify trace-context headers. The request is then considered as sampled.
```
Processing does not imply sampling.

yurishkuro replied:
"modify trace-context headers" - is "trace-context" defined somewhere as an umbrella term?

AloisReitbauer replied:
Trace context is the name of the spec

yurishkuro replied:
I'd remove the dash then, "modify the tracing context headers"

danielkhan replied:
Fixed

yurishkuro commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,47 @@
+5. The tracing system modifies the `traceparent` header. 
+    1. **Update `parent-id`**. The value of property `parent-id` can be set to the new value representing the ID of the current operation. This is the most common mutation and should be considered a default.
+    > **PENDING DECISION**: This definition is not entirely clear.   
+    2. **Indicate recorded state**. The value of `recorded` flag of `trace-flags` may be set to `1` if it had the value `0` before or vice versa. `parent-id` MUST be set to the new value with the `recorded` flag update. See details of `recorded` flag for more information on how this flag is recommended to be used.
```
There is no relation between mutating parentid and recorded flag. I think this para can be removed.

AloisReitbauer replied:
Agreed. But we still need to add language on changing recorded state. Would changing from enumeration to bulleted list do the trick for you?

yurishkuro replied:
perhaps the first point can use MUST and the second MAY.

danielkhan replied:
fixed in https://github.com/w3c/trace-context/pull/246/commits/ce6589d1e4962766a4de65e15f45227d6d337002

yurishkuro commented on spec/30-processing-model.md:
```diff
@@ -0,0 +1,47 @@
+
+## Processing Model in Forwarding Mode
+
+In forwarding mode, a tracing system decides to forward tracing information but does not modify the trace-context headers. The request is then considered as not sampled.
```
Forwarding does not imply no sampling. The system may still record data that enriches the current span.

AloisReitbauer replied:
I don't understand your comment. We defined this mode for tracers that explicitly do not change the trace. We also had the wording participating and not-participating for a while.

yurishkuro replied:
Concrete example: we used to use HAProxy as a routing layer. It did not alter the tracing headers, so according to definitions here it was in the forwarding mode. However, it was still checking the tracing headers, and if the sampling bit was on it would record certain timing events and submit them to the tracing backend, where they would be merged into the client-side span of the caller.

danielkhan replied:
Resolved in https://github.com/w3c/trace-context/pull/246/commits/4b61d5c458e64aeb581ee769e76c86f0a16dbbcd

danielkhan committed:
Improve grammar, spelling and structure of HTTP-Header Format and processing model (#227)

* Grammar, Spelling and wording

- Fixed grammar, spelling and edited wording for clarity
- Formatted pending decisions
- Added issue links

* Moved IF conditions to bullets instead of ordered lists

* Auto-fixed spelling and grammar

* Edited and restructured for more clarity

* Edited example in intro

* Intro example - titles

* Edited and restructured for more consistency

Rephrased the intro example.

* Markdown linting

* ignore vscode files

* Removing vscode specific file

* removed vscode folder

* Fix line lengths

* Improved wording wrt byte arrays / hex

* Fixed formatting

* Removed generic tracer paragraph

This seems to be an implementation detail that should likely not be part of the spec.

* Removed paragraph about reuse of trace ID for correlation.

* Added 'field' when mentioning HTTP-header

* Removed redundant explanation of OWS

* Reverted the removed paragraphs to allow further discussion

* Changed wording from list to ordered set

* Changed wording to set

* Removed 'opnionated' remark

* changed wording to field value

yurishkuro commented on .gitignore:
```diff
@@ -20,4 +20,7 @@ bin
+.swp
+
+# VSCode
+.vscode
```
this can go to `~/.gitignore_global` instead

danielkhan replied:
We already have a .gitignore in this repo and it already contains ignore rules for specific IDEs. Are you suggesting to get rid of all that rules?

yurishkuro commented on spec/20-HTTP_HEADER_FORMAT.md:
```diff
@@ -1,30 +1,41 @@
-# Trace context HTTP headers format
+# Trace context HTTP headers
```
nit:  doesn't seem like this change belongs to this PR

danielkhan committed:
[WIP] Editorial changes to HTTP Header Format (#242)

* removed OWS explanation #238

* Removed traceparent additional info #168 #234

* Removed mutation as they are now part of processing model. Closes #240

* (re) added OWS description. Closes #238

* Added 'the'

* Remove mentioning of span-id. Closes #236

* Trimmed down wording on randomness. #Closes 235

* Adapted wording #235

* Removed mention of 'generic tracer'. Closes #233.

danielkhan committed:
Merge from upstream

danielkhan committed:
Merge branch 'cr--edit-processing-model' of github.com:w3c/trace-context into cr--edit-processing-model
